join-chat-to-teams-meeting/app/src/main/java/com/contoso/acsquickstart/MainActivity.java

```diff
@@ -0,0 +1,302 @@
+    private String threadId;
+    private ChatThreadAsyncClient chatThreadAsyncClient;
+
+    // The list of ids corresponsding to messages which have already been processed
+    ArrayList<String> chatMessages = new ArrayList<>();
+    TextView callStatusBar;
+    TextView recordingStatusBar;
+
+    private CallAgent agent;
+    private Call call;
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.activity_main);
+
+        getAllPermissions();
+        createAgent();
+
+        Button joinMeetingButton = findViewById(R.id.join_meeting_button);
+        joinMeetingButton.setOnClickListener(l -> joinTeamsMeeting());
+
+        Button hangupButton = findViewById(R.id.hangup_button);
+        hangupButton.setOnClickListener(l -> leaveMeeting());
+
+        callStatusBar = findViewById(R.id.call_status_bar);
+        recordingStatusBar = findViewById(R.id.recording_status_bar);
+    }
+
+    /**
+     * Join Teams meeting
+     */
+    private void joinTeamsMeeting() {
+        if (UserToken.startsWith("<")) {
+            Toast.makeText(this, "Please enter token in source code", Toast.LENGTH_SHORT).show();
+            return;
+        }
+
+        EditText calleeIdView = findViewById(R.id.teams_meeting_link);
+        String meetingLink = calleeIdView.getText().toString();
+        if (meetingLink.isEmpty()) {
+            Toast.makeText(this, "Please enter Teams meeting link", Toast.LENGTH_SHORT).show();
+            return;
+        }
+
+        JoinCallOptions options = new JoinCallOptions();
+        TeamsMeetingLinkLocator teamsMeetingLinkLocator = new TeamsMeetingLinkLocator(meetingLink);
+
+        call = agent.join(
+                getApplicationContext(),
+                teamsMeetingLinkLocator,
+                options);
+        call.addOnStateChangedListener(p -> setCallStatus(call.getState().toString()));
+
+        EditText threadIdView = findViewById(R.id.teams_meeting_thread_id);
+        threadId = threadIdView.getText().toString();
+        // Initialize Chat Thread Client
+        chatThreadAsyncClient = new ChatThreadClientBuilder()
+                .endpoint(ResourceUrl)
+                .credential(new CommunicationTokenCredential(UserToken))
+                .chatThreadId(threadId)
+                .buildAsyncClient();
+        Button sendMessageButton = findViewById(R.id.send_message_button);
+        EditText messageBody = findViewById(R.id.message_body);
+        // Register the method for sending messages and toggle the visibility of chat components
+        sendMessageButton.setOnClickListener(l -> sendMessage());
+        sendMessageButton.setVisibility(View.VISIBLE);
+        messageBody.setVisibility(View.VISIBLE);
+
+        // Start the polling for chat messages immediately
+        handler.post(runnable);
+    }
+
+    private Handler handler = new Handler();
```

Comment:
Why Java, Kotlin could be much compressed.

But for a repeating timer, 
https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ScheduledExecutorService.html

Also a good idea to check "isFinishing" and "isDestroyed" on the activity before doing something in a thread.